data/hooks/diagnosis/21-web.py

```diff
@@ -4,58 +4,153 @@
+                except requests.exceptions.Timeout:
+                    yield dict(meta={"test": "hairpinning"},
+                               status="WARNING",
+                               summary="diagnosis_http_hairpinning_issue",
+                               details=["diagnosis_http_hairpinning_issue_details"])
+                except:
+                    # Well I dunno what to do if that's another exception
+                    # type... That'll most probably *not* be an hairpinning
+                    # issue but something else super weird ...
+                    pass
+
+    def test_http(self, domains, ipversions):
+
+        results = {}
+        for ipversion in ipversions:
             try:
-                r = requests.post('https://diagnosis.yunohost.org/check-http', json={'domain': domain, "nonce": nonce}, timeout=30)
-                if r.status_code not in [200, 400, 418]:
-                    raise Exception("Bad response from the server https://diagnosis.yunohost.org/check-http : %s - %s" % (str(r.status_code), r.content))
-                r = r.json()
-                if "status" not in r.keys():
-                    raise Exception("Bad syntax for response ? Raw json: %s" % str(r))
-                elif r["status"] == "error" and ("code" not in r.keys() or not r["code"].startswith("error_http_check_")):
-                    if "content" in r.keys():
-                        raise Exception(r["content"])
-                    else:
-                        raise Exception("Bad syntax for response ? Raw json: %s" % str(r))
+                r = Diagnoser.remote_diagnosis('check-http',
+                                               data={'domains': domains,
+                                                     "nonce": self.nonce},
+                                               ipversion=ipversion)
+                results[ipversion] = r["http"]
             except Exception as e:
                 raise YunohostError("diagnosis_http_could_not_diagnose", error=e)
 
-            if r["status"] == "ok":
+        for domain in domains:
+
+            # If both IPv4 and IPv6 (if applicable) are good
+            if all(results[ipversion][domain]["status"] == "ok" for ipversion in ipversions):
+                if 4 in ipversions:
+                    self.do_hairpinning_test = True
                 yield dict(meta={"domain": domain},
                            status="SUCCESS",
-                           summary=("diagnosis_http_ok", {"domain": domain}))
-            else:
-                detail = r["code"].replace("error_http_check", "diagnosis_http") if "code" in r else "diagnosis_http_unknown_error"
+                           summary="diagnosis_http_ok")
+            # If both IPv4 and IPv6 (if applicable) are failed
+            elif all(results[ipversion][domain]["status"] != "ok" for ipversion in ipversions):
+                detail = results[4 if 4 in ipversions else 6][domain]["status"]
```

Comment:
Why not both ipversions here?

Reply:
Because I was thinking that it's probably the same error for both ip versions and we don't want to display the message twice ... But yea that's a bit weird :s